runtimes/huggingface/mlserver_huggingface/codecs.py

```diff
@@ -1,89 +1,351 @@
-from typing import Optional, Type, Any, Dict
+import json
+from typing import Optional, Type, Any, Dict, List, Union
+from functools import partial
+from xmlrpc.client import Boolean
 from mlserver.codecs.utils import (
+    InputOrOutput,
     has_decoded,
     _save_decoded,
     get_decoded_or_raw,
+    is_single_value,
+    set_content_type,
 )
+
+from mlserver.codecs.lists import is_list_of
 from mlserver.codecs.base import (
     RequestCodec,
     register_request_codec,
+    register_input_codec,
+    find_input_codec,
+    find_input_codec_by_payload,
     InputCodec as InputCodecTy,
 )
-from mlserver.codecs import StringCodec
+from mlserver.codecs import (
+    InputCodec,
+    StringCodec,
+)
+from mlserver.codecs.base64 import _decode_base64, _encode_base64
+from mlserver.codecs.lists import as_list
+from mlserver.types import RequestInput, ResponseOutput
+from PIL import Image
+
 from mlserver.types import (
     InferenceRequest,
     InferenceResponse,
 )
+from transformers.pipelines import Conversation
+
+from mlserver.types.dataplane import Parameters
+from .common import (
+    CommonJSONEncoder,
+    CONTENT_TYPE_JSON,
+    CONTENT_TYPE_IMAGE,
+    CONTENT_TYPE_CONVERSATION,
+    deserialize_conversation,
+    get_img_bytes,
+    build_image,
+    metainfo_for,
+)
 
 
-class MultiInputRequestCodec(RequestCodec):
+def get_input_codec(item: InputOrOutput, default: Optional[Type[InputCodecTy]]):
+    if item.parameters is None:
+        return default
+    if not item.parameters.content_type:
+        return default
+    return find_input_codec(item.parameters.content_type)
+
+
+def find_input_codec_by_payload_content(data: Any):
     """
-    The SingleInputRequestCodec can be used as a "meta-implementation" for other
-    codecs. Its goal to decode the whole request simply as the first decoded
-    element.
+    for huggingface JSONCodec has high priority
     """
+    if JSONCodec.can_encode(data):
+        return JSONCodec
+    if ImageCodec.can_encode(data):
+        return ImageCodec
+    if ConversationCodec.can_encode(data):
+        return ConversationCodec
+    return find_input_codec_by_payload(data)
```

Comment:
If we decorate them with `@register_input_codec` or `@register_request_codec` we should then be able to find them with `find_input_codec(payload=data)` (as in, they'll go into the general codec registry, which is not a bad thing).

Reply:
Some data may encode by multiple codecs, but there is no priority with codecs, `mlserver.codecs._find_codec_by_payload` just return the first codecs matched; I want to find codecs in order with priority, and that's the reason; 

If my understanding of the encoder registry is wrong, please let me know

Reply:
Oh, I see. Ok, that makes sense. In that case, it may be worth dropping a comment on that method to briefly explain that reasoning (i.e. and avoid people changing that in the future).